lib/WasmReader/WasmBinaryReader.cpp

```diff
@@ -530,39 +546,63 @@ WasmBinaryReader::FunctionHeader()
     sig = m_moduleInfo->GetSignature(sigId);
     m_funcInfo->SetSignature(sig);
 
+    TRACE_WASM_DECODER(L"Function header: flags = %x, sig = %u, size = %u", flags, sigId, m_funcState.size);
+}
+
+void
+WasmBinaryReader::FunctionBodyHeader()
+{
+    UINT i32Count = 0, i64Count = 0, f32Count = 0, f64Count = 0;
+    UINT len = 0;
+    UINT entryCount = LEB128(len);
+
     // locals
-    if (flags & bFuncDeclLocals)
+    for (UINT i = 0; i < entryCount; i++)
     {
-        // has locals
-        i32Count = ReadConst<UINT16>();
-        i64Count = ReadConst<UINT16>();
-        f32Count = ReadConst<UINT16>();
-        f64Count = ReadConst<UINT16>();
-
-        for (UINT i = 0; i < i32Count; i ++)
-        {
-            m_funcInfo->AddLocal(Wasm::WasmTypes::I32);
-        }
-        for (UINT i = 0; i < i64Count; i++)
+        UINT count = LEB128(len);
+        UINT8 type = ReadConst<UINT8>();
+        switch (type)
         {
-            m_funcInfo->AddLocal(Wasm::WasmTypes::I64);
-        }
-        for (UINT i = 0; i < f32Count; i++)
-        {
-            m_funcInfo->AddLocal(Wasm::WasmTypes::F32);
-        }
-        for (UINT i = 0; i < f64Count; i++)
-        {
-            m_funcInfo->AddLocal(Wasm::WasmTypes::F64);
+        case Wasm::WasmTypes::WasmType::I32:
+            i32Count += count;
+            break;
+        case Wasm::WasmTypes::WasmType::I64:
+            i64Count += count;
+            break;
+        case Wasm::WasmTypes::WasmType::F32:
+            f32Count += count;
+            break;
+        case Wasm::WasmTypes::WasmType::F64:
+            f64Count += count;
+            break;
+        default:
+            ThrowDecodingError(L"Unexpected local type");
         }
     }
 
+    for (UINT i = 0; i < i32Count; i++)
+    {
+        m_funcInfo->AddLocal(Wasm::WasmTypes::I32);
+    }
```

Comment:
Maybe I am missing something, but shouldn't' we have a list of funcInfos signatures contructed from FuncSigs that we attribute function bodies to here ? 
See [Luke's PR](https://github.com/WebAssembly/design/blob/0d96ed985368fc636a38de31970dc35274018ada/BinaryEncoding.md#function-bodies-section)